File: 0348-design-tic-tac-toe/0348-design-tic-tac-toe.py
class TicTacToe:

    def __init__(self, n: int):
        self.rows: List[int] = [0] * n
        self.cols: List[int] = [0] * n
        self.diagonal: int = 0
        self.antiDiagonal: int = 0
        self.n = n
        self.board: List[List[int]] = [[0 for _ in range(n)] for _ in range(n)]

    def isValidMove(self, row: int, col: int) -> bool:
        if not (0 <= row < self.n and 0 <= col < self.n):
            raise IndexError("Move out of bounds!")
        if self.board[row][col] != 0:
            raise ValueError(f"Cell ({row}, {col}) is already taken")

    def move(self, row: int, col: int, player: int) -> int:

        self.isValidMove(row, col)

        self.board[row][col] = player

        currentPlayer: int = 1 if player == 1 else -1
        self.rows[row] += currentPlayer
        self.cols[col] += currentPlayer

        if row == col:
            self.diagonal += currentPlayer
        
        if col == (len(self.cols) - row - 1):
            self.antiDiagonal += currentPlayer

        if (abs(self.rows[row]) == self.n or
            abs(self.cols[col]) == self.n or
            abs(self.diagonal) == self.n or
            abs(self.antiDiagonal) == self.n):
            return player
        return 0
        


# Your TicTacToe object will be instantiated and called as such:
# obj = TicTacToe(n)
# param_1 = obj.move(row,col,player)

File: 0348-design-tic-tac-toe/0348-design-tic-tac-toe.java
import java.util.Arrays;

public class TicTacToe {
    private final int n;
    private final int[] rows;
    private final int[] cols;
    private int diagonal;
    private int antiDiagonal;
    private final int[][] board;

    public TicTacToe(int n) {
        this.n = n;
        this.rows = new int[n];
        this.cols = new int[n];
        this.diagonal = 0;
        this.antiDiagonal = 0;
        this.board = new int[n][n];          // all zeroes by default
    }

    private boolean isValid(int row, int col) {
        return row >= 0 && row < n && col >= 0 && col < n;
    }

    /**
     * Player {1, 2} makes a move at (row, col).
     * @return   0 = no one won yet, 1 = player 1 wins, 2 = player 2 wins, -1 = invalid move
     */
    public int move(int row, int col, int player) {
        // out of bounds or cell already taken
        if (!isValid(row, col) || board[row][col] != 0) {
            return -1;
        }

        // mark the board
        board[row][col] = player;

        // convert player to +1 / -1
        int toAdd = (player == 1 ? +1 : -1);

        // update row/col counters
        rows[row] += toAdd;
        cols[col] += toAdd;

        // update diagonals if applicable
        if (row == col) {
            diagonal += toAdd;
        }
        if (col == (n - row - 1)) {
            antiDiagonal += toAdd;
        }

        // check if this move causes a win
        if (Math.abs(rows[row]) == n
         || Math.abs(cols[col]) == n
         || Math.abs(diagonal) == n
         || Math.abs(antiDiagonal) == n) {
            return player;
        }

        // no winner yet
        return 0;
    }
}
